Add driver for TPMv2

ruchi393 opened the pull request:
This PR supersedes the earlier PR https://github.com/OP-TEE/optee_os/pull/5090

Changes from the earlier pR
- Code has been restructured to introduce a layer for FIFO interface based on PTP (Platform TPM Profile) specification.
- Send and recieve code has been changed to follow the Device Driver Design principles for TPM 2.0
- Command structure along with commands to selftest and startup TPM have been added

The driver can be tested on QEMU platform with swtpm interface.

This is the first PR for this work. Next steps for this work include
- adding commands to read and extend PCR's in TPM
- Add a SPI based backend for the TPM2

<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         https://optee.readthedocs.io/en/latest/general/contribute.html

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" in the contribution guidelines.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc. More details
       about this can also be found on the link provided above.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

ruchi393 commented:
Fixed checkpatch warning about not have a matching Signed-off for Co-developed-by.

ruchi393 commented:
Updated the PR with following changes
- Issue in addition of more than 1 commands
- selftest and startup call moved to tpm_chip

ruchi393 commented:
Rebased to resolve conflicts with master

@apalos , @etienne-lms , @jforissier , @jenswi-linaro , can you please help review.

jenswi-linaro commented on core/include/drivers/tpm2/tpm2_cmd.h:
```diff
@@ -0,0 +1,44 @@
+#include <io.h>
+#include <stdint.h>
+
+struct tpm2_cmd_hdr {
```
These structs looks a bit scary. It seems that they are here just for the inline functions below and apart from that only internal us in `tpm2_cmd.c`. A comment mentioning how this these structs are supposed to be used would be welcome.

ruchi393 replied:
This is the command header for TPM related commands. I will add comments to describe it more
It's mostly used in tpm2_cmd.c where tpm commands are created. It's also used in tpm2_ptp_fifo.c when tpm needs to send command or receive response.

jenswi-linaro commented on core/include/drivers/tpm2/tpm2_cmd.h:
```diff
@@ -0,0 +1,44 @@
+struct tpm2_cmd_hdr {
+	uint16_t tag;
+	uint32_t size;
+	union {
```
Isn't this overkill? `uint32_t code;` could be used instead, it's only supposed to be accessed by special functions anyway.

ruchi393 replied:
Ok..TPM has lots of structures...I kept a union just to avoid confusion..But since we are accessing it via special functions I will remove the union

jenswi-linaro commented on core/drivers/tpm2/tpm2_ptp_fifo.c:
```diff
@@ -0,0 +1,438 @@
+	uint8_t access = 0;
+
+	do {
+		rc = ops->rx8(chip, TPM2_ACCESS(loc), 1, &access);
```
s/1/sizeof(access)/

jenswi-linaro commented on core/drivers/tpm2/tpm2_ptp_fifo.c:
```diff
@@ -0,0 +1,438 @@
+	uint8_t locality = 0;
+	int rc = -1;
+
+	rc = ops->rx8(chip, TPM2_ACCESS(loc), 1, &locality);
```
s/1/sizeof(access)/

jenswi-linaro commented on core/drivers/tpm2/tpm2_ptp_fifo.c:
```diff
@@ -0,0 +1,438 @@
+	return TPM2_ERR_TIMEOUT;
+}
+
+static bool tpm2_fifo_check_locality(struct tpm2_chip *chip, uint32_t loc)
```
It seems this function does more than the name suggests. A comment somewhere on what's going on would be welcome.

jenswi-linaro reviewed (commented):
Some early comments on "drivers/tpm2: Add basic TPM2 support in OP-TEE"

ldts commented:
@ruchi393 the SE05x can implement most of the TPM2 functionality - currently looking into PCR.

The interface would however be different to what is done with tpm2_mmio.c: the SE05x driver will just implement the tpm2_* functions. But we will need to access your follow up work below

```
* Parse the eventlog passed to OP-TEE and extend the PCR's
```

just sharing this use case to let you know how I see this PR evolving and seeing if you have any thoughts for maybe a higher level API? Perhaps the SE05x could register itself as a tpm2 provider to tpm2_cmd.c? and the current implementation of tpm2_cmd using the chip move to a separate file compiled conditionally?

Then the SE05x TPM2 implementation (maybe CAAM or other cryptoICs will do others) would fit nicely in your design I think.

ruchi393 commented on core/include/drivers/tpm2/tpm2_cmd.h:
```diff
@@ -0,0 +1,44 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
```
Won't moving all tpm2* files to drivers directory clutter things a little. I had created a separate directory to have all the tpm related files at one place.

ruchi393 replied:
Have moved all header files out of tpm2 dir. However I have still kept the C files still in core/drivers/tpm2 folder. Do let me know if this is fine.

jenswi-linaro replied:
Sound good to me.

ruchi393 commented on core/include/drivers/tpm2/tpm2_chip.h:
```diff
@@ -0,0 +1,99 @@
+	TPM2_ERR_IO = 6,
+	TPM2_ERR_NODEV	= 7,
+
+	/* For command errors, cmd error will be OR'ed with TPM2_ERR_CMD */
```
You are right. Initially I was thinking of adding the command errors here. I will remove the comment

ruchi393 commented on core/include/tpm2.h:
```diff
@@ -0,0 +1,43 @@
+#include <util.h>
+
+/* TPM2_ST Structure Tags */
+typedef uint16_t TPM2_ST;
```
As I mentioned TPM has a lot of structures, constants and defines. These come from TPM specifications.
Here I was just trying to be as much consistent as possible with their definition from this guide
TCG TSS 2.0 Overview and Common Structures Specification - 
https://trustedcomputinggroup.org/wp-content/uploads/TSS_Overview_Common_v1_r10_pub09232021.pdf

ruchi393 replied:
I have cleaned the file up

ruchi393 commented on core/drivers/tpm2/tpm2_mmio.c:
```diff
@@ -0,0 +1,107 @@
+	.recv = tpm2_fifo_recv,
+};
+
+static struct tpm2_chip tpm2_mmio_chip = {
```
can't be const, as lot of the variables in it would get initialised as tpm chip gets registered

jforissier replied:
Right, my mistake

ruchi393 commented on core/arch/arm/plat-vexpress/main.c:
```diff
@@ -24,12 +24,19 @@
 #include <sm/psci.h>
 #include <stdint.h>
 #include <string.h>
+#if defined(PLATFORM_FLAVOR_qemu_armv8a) && defined(CFG_TPM2_MMIO)
```
I will take care of this when I squash the patches

ruchi393 commented:
> @ruchi393 the SE05x can implement most of the TPM2 functionality - currently looking into PCR.
> 
> The interface would however be different to what is done with tpm2_mmio.c: the SE05x driver will just implement the tpm2_* functions. But we will need to access your follow up work below
> 
> ```
> * Parse the eventlog passed to OP-TEE and extend the PCR's
> ```
> 
> just sharing this use case to let you know how I see this PR evolving and seeing if you have any thoughts for maybe a higher level API? Perhaps the SE05x could register itself as a tpm2 provider to tpm2_cmd.c? and the current implementation of tpm2_cmd using the chip move to a separate file compiled conditionally?
> 
> Then the SE05x TPM2 implementation (maybe CAAM or other cryptoICs will do others) would fit nicely in your design I think.

Thanks @ldts for describing the usecase. I will keep this in mind when creating calls to extend pcrs.

ruchi393 commented:
@jenswi-linaro , @jforissier , PR updated with fixes based on comments till date.

etienne-lms commented on core/drivers/tpm2/tpm2_chip.c:
```diff
@@ -0,0 +1,64 @@
+
+enum tpm2_result tpm2_chip_send(uint8_t *buf, uint32_t len)
+{
+	if (!tpm2_device || !tpm2_device->ops)
```
`tpm2_chip_register()` already checks `tpm2_device->ops != NULL`.
Maybe it should verify `.send` and `.recv` are populated?

etienne-lms commented on core/drivers/tpm2/tpm2_chip.c:
```diff
@@ -0,0 +1,64 @@
+		return TPM2_ERR_NODEV;
+	}
+
+	return ret;
```
return TPM2_OK;

etienne-lms commented on core/drivers/tpm2/tpm2_mmio.c:
```diff
@@ -0,0 +1,92 @@
+
+	base.pa = pbase;
+
+	base.va = io_pa_or_va_secure(&base, TPM2_REG_SIZE);
```
`io_pa_or_va_secure(&base, ...)` already loads `base.va`.
For consistency I suggest:
```c
	tpm2_mmio_base = io_pa_or_va_secure(&base, TPM2_REG_SIZE);
	assert(tpm2_mmio_base);

etienne-lms commented on core/drivers/tpm2/tpm2_mmio.c:
```diff
@@ -0,0 +1,92 @@
+
+	tpm2_mmio_base = base.va;
+
+	DMSG("TPM2 MMIO pbase: 0x%" PRIxVA, base.pa);
```
PRIxPA

etienne-lms commented on core/drivers/tpm2/tpm2_ptp_fifo.c:
```diff
@@ -0,0 +1,463 @@
+
+	chip->locality = -1;
+
+	return ret;
```
return TPM2_OK;

etienne-lms commented on core/drivers/tpm2/tpm2_ptp_fifo.c:
```diff
@@ -0,0 +1,463 @@
+	if (chip->locality < 0)
+		return TPM2_ERR_INVALID_ARG;
+
+	rc = ops->tx8(chip, TPM2_STS(chip->locality), sizeof(buf), &buf);
```
`return ops->tx8(...);`

etienne-lms commented on core/drivers/tpm2/tpm2_ptp_fifo.c:
```diff
@@ -0,0 +1,463 @@
+
+	rc = ops->rx8(chip, TPM2_STS(chip->locality), sizeof(*status), status);
+	if (rc < 0)
+		return TPM2_ERR_IO;
```
`return rc;`

etienne-lms commented on core/drivers/tpm2/tpm2_ptp_fifo.c:
```diff
@@ -0,0 +1,463 @@
+	if (rc < 0)
+		return TPM2_ERR_IO;
+
+	/* Few bits in STS register are alays expected to be 0. */
```
s/alays/always/

etienne-lms commented on core/drivers/tpm2/tpm2_ptp_fifo.c:
```diff
@@ -0,0 +1,463 @@
+/* Different status bit settings can require different timeouts */
+static enum tpm2_result tpm2_fifo_wait_for_status(struct tpm2_chip *chip,
+						  uint8_t mask,
+						  uint32_t timeout,
```
i suggest to rename `timeout` to `timeout_ms`

etienne-lms commented on core/drivers/tpm2/tpm2_ptp_fifo.c:
```diff
@@ -0,0 +1,463 @@
+	timeout_ref = timeout_init_us(chip->timeout_a * 1000);
+	/* wait for burstcount */
+	do {
+		ops->rx32(chip, TPM2_STS(chip->locality), &burst);
```
check the return value.

etienne-lms commented on core/drivers/tpm2/tpm2_ptp_fifo.c:
```diff
@@ -0,0 +1,463 @@
+{
+	/* Cancel any command that may have been sent to TPM */
+	tpm2_fifo_set_status(chip, TPM2_STS_COMMAND_READY);
+	/* Relinquish locality if it waas requested earlier */
```
s/waas/was/

etienne-lms commented on core/drivers/tpm2/tpm2_ptp_fifo.c:
```diff
@@ -0,0 +1,463 @@
+			ret = tpm2_fifo_get_burstcount(chip, &burstcnt);
+			if (ret) {
+				retry_cnt++;
+				continue;
```
this while continue in the current `while()` loop, not the one above, if i'm right.

ruchi393 replied:
Good catch. Thanks. Will fix this

etienne-lms commented on core/drivers/tpm2/tpm2_ptp_fifo.c:
```diff
@@ -0,0 +1,463 @@
+				param_size_flag = 1;
+
+				if (bytes2read > *len) {
+					EMSG("Buffer too small: %x > %x",
```
PRIx32 for both

etienne-lms commented on core/include/tpm2.h:
```diff
@@ -0,0 +1,39 @@
+#define	TPM2_CC_PCR_EXTEND	U(0x00000182)
+
+/* Command Error Codes */
+#define TPM2_RC_SUCCESS		0x0
```
Seems unused.

ruchi393 replied:
Ok will remove it.

etienne-lms commented on core/include/tpm2.h:
```diff
@@ -0,0 +1,39 @@
+#define TPM2_RC_SUCCESS		0x0
+
+int tpm2_startup(uint16_t mode);
+int tpm2_selftest(uint8_t full);
```
Could use `enum tpm2_result` or `TEE_Result` error code.

ruchi393 replied:
I am using int as want to distinguish between tpm interface errors and command related errors.. Have added a description for these commands...Let me know if that looks ok..

etienne-lms replied:
Looks a bit strange. This adds a specific positive TPM2 error core, and negative... system error code.
I think this function should preferably return either a `enum tpm2_result` or `TEE_Result` to avoid confusion.
If `enum tpm2_result`: `TPM2_ERR_GENERIC` could match any drive rissue (out-of-mem etc..).
If `TEE_Result`: `TEE_ERROR_COMMUNICATION` could match any TPM2 comm failures.

I think `tpm2_startup()` should also return either a `TEE_Result` or a `enum tpm2_result`

ruchi393 replied:
Changing it to tpm2_result.

I am also rewriting these functions to always pass a command and response buffer. These buffers can be on stack for startup and selftest as they are going to be small

etienne-lms commented on core/include/drivers/tpm2_cmd.h:
```diff
@@ -0,0 +1,49 @@
+
+#include <compiler.h>
+#include <io.h>
+#include <stdint.h>
```
`#include <utee_defines.h>` for `TEE_U32_FROM_BIG_ENDIAN()`

ruchi393 commented on core/drivers/tpm2/tpm2_cmd.c:
```diff
@@ -0,0 +1,173 @@
+	int ret = -1;
+
+	if (cmd_len > bufsz)
+		return ret;
```
This function can return 2 types of errors
- TPM interface related errors (returned by TPM driver - TPM2_ERR)
- TPM command related errors. Retrieved from response buffer.

I have been thinking that perhaps use int and define the errors 
= 0 SUCCESS
< 0 [ TPM interface errors -> I can perhaps return a -TPM_ERR_ here)
>0 [ These will be command errors from response buffer ]

What do you think ? Any suggestions are welcome.

ruchi393 commented:
PR updated ..all comments  taken care of

etienne-lms commented on core/include/tpm2.h:
```diff
@@ -0,0 +1,56 @@
+/*
+ * Send a TPM2_Startup command
+ *
+ * @mode - TPM startup mode
```
state that @mode is one of `TPM2_SU_CLEAR`|`TPM2_SU_STATE`?

etienne-lms commented on core/drivers/tpm2/tpm2_ptp_fifo.c:
```diff
@@ -0,0 +1,477 @@
+		sent = 0;
+		wr_size = 0;
+		burstcnt = 0;
+		status = 0;
```
nitpick: i would suggest to move `buf_len = len;` and `sent = 0;` to line 313, right before the loop that uses them. Initialization of `wr_size`, `burstcnt`, `status` are not needed.

ruchi393 replied:
Thanks. You are right. I will make the change

etienne-lms commented on core/drivers/tpm2/tpm2_ptp_fifo.c:
```diff
@@ -0,0 +1,477 @@
+	if (*len < TPM2_HDR_LEN) {
+		EMSG("Unable to read TPM2 header\n");
+		*len = TPM2_HDR_LEN;
+		return TPM2_ERR_INVALID_ARG;
```
An inline comment should describe that output `len` is updated on `TPM2_ERR_INVALID_ARG` .
I even think a specific error code is needed here, to distinguish from other places where the function may return `TPM2_ERR_INVALID_ARG`.

ruchi393 replied:
Ok..Will add a new error code TPM2_ERR_SHORT_BUFFER.

etienne-lms commented on core/drivers/tpm2/tpm2_ptp_fifo.c:
```diff
@@ -0,0 +1,477 @@
+				bytes2read = tpm2_cmd_len((void *)buf);
+				param_size_flag = 1;
+
+				if (bytes2read > *len) {
```
`bytes2read + TPM2_HDR_LEN > *len`

etienne-lms replied:
ok

ruchi393 commented:
Thanks a lot for the review @etienne-lms . I have addressed your comments and updated the PR

ruchi393 commented:
@etienne-lms , @jenswi-linaro , PR has been updated with comments addressed

ruchi393 commented on core/include/drivers/tpm2_chip.h:
```diff
@@ -0,0 +1,98 @@
+ * Based on:
+ * linux/errno.h (very loosely)
+ */
+enum tpm2_result {
```
These are needed for TPM interface functions to distinguish between device related errors

jenswi-linaro replied:
OK, please replace the comment above with this information.

ruchi393 commented on core/drivers/tpm2/tpm2_cmd.c:
```diff
@@ -0,0 +1,155 @@
+		return -1;
+
+	put_be16(&cmd->hdr.tag, tag);
+	put_be32(&cmd->hdr.size, sizeof(struct tpm2_cmd_hdr));
```
Sure..will change it

ruchi393 commented on core/drivers/tpm2/tpm2_cmd.c:
```diff
@@ -0,0 +1,155 @@
+
+	rc = tpm2_transmit(buf, tpm2_cmd_len((struct tpm2_cmd *)buf), resp_buf,
+			   &resp_len);
+	if (rc && rc == TPM2_ERR_SHORT_BUFFER)
```
Yes, rc checking not needed ...will remove

ruchi393 commented:
PR updated with review comments addressed. @jenswi-linaro , @etienne-lms please have a look

etienne-lms reviewed (commented):
late comments....

jenswi-linaro commented:
For "core: io: add {get/put}_unaligned_be{16/32/64}()" please apply:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

The rest are:
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

ruchi393 commented on core/drivers/tpm2/tpm2_chip.c:
```diff
@@ -0,0 +1,64 @@
+		EMSG("TPM2 Self Test Failed");
+	}
+
+	rc = tpm2_startup(TPM2_SU_CLEAR);
```
My bad. Sorry, forgot to fix this

ruchi393 commented:
Thanks @etienne-lms and @jenswi-linaro . I have updated the PR.

Took care of last set of comments. Rebased patches over master, squashed them and added tags from @jenswi-linaro

etienne-lms commented on core/drivers/tpm2/tpm2_chip.c:
```diff
@@ -0,0 +1,61 @@
+#include <tpm2.h>
+#include <trace.h>
+
+struct tpm2_chip *tpm2_device;
```
could be `static`

etienne-lms commented on core/drivers/tpm2/tpm2_cmd.c:
```diff
@@ -0,0 +1,156 @@
+		return TPM2_CMD_DURATION_DEFAULT;
+	}
+
+	return TPM2_CMD_DURATION_DEFAULT;
```
remove this line

etienne-lms commented on core/drivers/tpm2/tpm2_cmd.c:
```diff
@@ -0,0 +1,156 @@
+	enum tpm2_result rc = TPM2_OK;
+	uint32_t err_code = 0;
+
+	if (!buf || !bufsz || !resp || !len || (len && !*len))
```
`!bufsz` is already test below

etienne-lms commented on core/drivers/tpm2/tpm2_cmd.c:
```diff
@@ -0,0 +1,156 @@
+	uint32_t resp_len = sizeof(resp_buf);
+	enum tpm2_result rc = TPM2_OK;
+
+	tpm2_cmd_init_hdr(buf, buf_len, TPM2_ST_NO_SESSIONS, TPM2_CC_STARTUP);
```
test value returned.
ditto at line 127, 146 and 147.

By the way, I think `tpm2_cmd_init_hdr()` and `tpm2_cmd_add_*()` functions should be changed to return a `enum tpm2_result`. Makes tpm2/ implementation more consistent.

ruchi393 replied:
I was in double mind over here whether to add a check at every step or not. The only error condition is if buffer allocated is not enough. The command writer/caller using these commands should have allocated enough buffer size based on the specification. Even if he makes an error and we continue, cmd TPM receives will be mal-formed and will result in a cmd error. This should be taken care of in development stage itself.
I have no specific preference but adding a check with every buffer addition would add too many checks as I add more commands which are more complex.

ruchi393 replied:
I have added the change with a panic if the return is an error. Please have a look.

etienne-lms commented on core/drivers/tpm2/tpm2_mmio.c:
```diff
@@ -0,0 +1,90 @@
+		return ret;
+	}
+
+	DMSG("Init and open ok");
```
could remove this trace message

etienne-lms commented on core/drivers/tpm2/tpm2_ptp_fifo.c:
```diff
@@ -0,0 +1,484 @@
+enum tpm2_result tpm2_fifo_end(struct tpm2_chip *chip)
+{
+	/* Cancel any command that may have been sent to TPM */
+	tpm2_fifo_set_status(chip, TPM2_STS_COMMAND_READY);
```
Should this function consider that these 2 function may fail?

etienne-lms commented on core/drivers/tpm2/tpm2_ptp_fifo.c:
```diff
@@ -0,0 +1,484 @@
+		return TPM2_OK;
+	}
+
+	/* This will result in current command getting cancelled */
```
could call `tpm2_fifo_end(chip)`

etienne-lms commented on core/drivers/tpm2/tpm2_ptp_fifo.c:
```diff
@@ -0,0 +1,484 @@
+		}
+
+		rxsize = 0;
+		/* Initialy read the TPM header */
```
s/Initialy/Initially/
or s/Initialy/First/

etienne-lms commented on core/drivers/tpm2/tpm2_ptp_fifo.c:
```diff
@@ -0,0 +1,484 @@
+	}
+out:
+	/* This will result in current command getting cancelled */
+	tpm2_fifo_set_status(chip, TPM2_STS_COMMAND_READY);
```
could call `tpm2_fifo_end(chip)`

etienne-lms reviewed (commented):
still minor comments and missing tests on function return values in `tpm2_startup()`.

ruchi393 commented:
PR updated with fixes based on comments

etienne-lms commented on core/drivers/tpm2/tpm2_ptp_fifo.c:
```diff
@@ -369,10 +372,8 @@ enum tpm2_result tpm2_fifo_send(struct tpm2_chip *chip, uint8_t *buf,
-	/* We are here because some error occurred, so relinquish locality */
-	tpm2_fifo_relinquish_locality(chip);
+	/* Cancel command and relinquish locality */
+	ret = tpm2_fifo_end(chip);
```
sorry I confused you here.
Sorry not overwrite returned `ret` value.
I would propose:
```c
	/* Cancel command and relinquish locality */
	if (tpm2_fifo_end(chip))
		return TPM2_ERR_GENERIC;

	return ret;
```
Ditto at line 480

etienne-lms reviewed (commented):
One remaining issue with an error case in `tpm2_fifo_send()` (commit "drivers/tpm2: Add basic TPM2 support in OP-TEE").
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` for the series once addressed.

ruchi393 commented:
@etienne-lms , I have added an RFC patch to  change the prototype of functions adding tpm commands. The caller is supposed to have enough buffer space allocated based on the specifications. If not this should panic. So  have moved the panic to base function rather than every function which creates command.

Let me know what you think.

etienne-lms commented:
Appended RFC commit looks good to me. You can apply my A-b tag.
Maybe even `if (len < req_len) panic();` could be replaced with `assert(len >= req_len);`

ruchi393 commented:
> Appended RFC commit looks good to me. You can apply my A-b tag. Maybe even `if (len < req_len) panic();` could be replaced with `assert(len >= req_len);`

Done. Thanks @etienne-lms

ruchi393 commented:
Squashed the RFC commit. Tags applied.
